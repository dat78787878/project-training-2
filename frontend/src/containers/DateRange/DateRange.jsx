import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import React from 'react';
import { useDispatch } from 'react-redux';
import { useState, useEffect, useCallback } from 'react';
import { Button } from 'react-bootstrap';
import { getData } from '../../redux/dateRange/actions';
import moment from 'moment';
import { Link } from 'react-router-dom';

const DateRange = () => {
  const dispatch = useDispatch();
  const [startDate, setStartDate] = useState(moment(new Date()));
  const [endDate, setEndDate] = useState(moment(new Date()));
  const [start, setStart] = useState();
  const [end, setEnd] = useState();
  const [focusedInput, setFocusedInput] = useState(null);
  const handleDatesChange = ({ startDate, endDate }) => {
    setStartDate(startDate);
    setEndDate(endDate);
    setStart(moment(startDate).format('YYYY-MM-DD'));
    setEnd(moment(endDate).format('YYYY-MM-DD'));
  };
  useEffect(() => {
    const urlSearchParams = new URLSearchParams(window.location.search);
    const params = Object.fromEntries(urlSearchParams.entries());
    setStartDate(moment(params.fromDate_));
    setEndDate(moment(params.toDate_));
  }, [window.location.search]);

  const handleClick = useCallback(() => {
    dispatch(
      getData({
        fromDate: start,
        toDate: end
      })
    );
  }, [startDate, endDate]);

  return (
    <div className="dateRange d-flex">
      <div className="dateRange-title">Range</div>
      <DateRangePicker
        startDate={startDate}
        startDateId="start-date"
        endDate={endDate}
        endDateId="end-date"
        onDatesChange={handleDatesChange}
        focusedInput={focusedInput}
        onFocusChange={(focusedInput) => setFocusedInput(focusedInput)}
      />
      <Link to={`?fromDate_=${start}&toDate_=${end}`}>
        <Button variant="success" onClick={handleClick} className="m-1">
          OKE
        </Button>
      </Link>
    </div>
  );
};

export default DateRange;
