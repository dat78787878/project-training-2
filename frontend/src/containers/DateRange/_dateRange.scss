.dateRange {
  padding-top: 10px;
  padding-left: 5px;
  &-title {
    padding: 10px;
  }
}
