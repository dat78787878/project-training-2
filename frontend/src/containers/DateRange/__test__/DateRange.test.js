import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DateRange from '../DateRange';
import { useDispatch } from 'react-redux';
import 'react-dates/initialize';
import { BrowserRouter } from 'react-router-dom';
import moment from 'moment';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn()
}));

describe('DateRange', () => {
  const mockDispatch = jest.fn();
  beforeEach(() => {
    useDispatch.mockImplementation(() => mockDispatch);
  });
  afterEach(() => {
    useDispatch.mockClear();
  });
  it('renders correctly', () => {
    const { container } = render(
      <BrowserRouter>
        <DateRange />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
    expect(container).toBeTruthy();
    expect(container.getElementsByClassName('dateRange'));
  });
  it('change date range and dispatch', () => {
    const mockDate = new Date();
    const buttonOK = getByText('OKE');
    const { container, getByRole, getByText } = render(
      <BrowserRouter>
        <DateRange />
      </BrowserRouter>
    );
    getByRole('textbox', { name: /Start Date/i }).focus();
    expect(getByRole('textbox', { name: /Start Date/i }).value).toEqual(
      moment(mockDate).format('MM/DD/YYYY')
    );
    fireEvent.click(buttonOK);
    expect(container).toMatchSnapshot();
  });
});
