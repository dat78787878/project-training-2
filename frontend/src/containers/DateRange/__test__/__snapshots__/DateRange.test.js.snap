// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange change date range and dispatch 1`] = `
<div>
  <div
    class="dateRange d-flex p-2"
  >
    <div
      class="row"
    >
      <div
        class="col-md-2 col-sm-12"
      >
        <div
          class="dateRange-title"
        >
          Range
        </div>
      </div>
      <div
        class="col-md-8 col-sm-12"
      >
        <div
          class="DateRangePicker DateRangePicker_1"
        >
          <div>
            <div
              class="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__withBorder DateRangePickerInput__withBorder_2"
            >
              <div
                class="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-start-date"
                  aria-label="Start Date"
                  autocomplete="off"
                  class="DateInput_input DateInput_input_1"
                  id="start-date"
                  name="start-date"
                  placeholder="Start Date"
                  type="text"
                  value="04/03/2022"
                />
                <p
                  class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-start-date"
                >
                  Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
              <div
                aria-hidden="true"
                class="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
                role="presentation"
              >
                <svg
                  class="DateRangePickerInput_arrow_svg DateRangePickerInput_arrow_svg_1"
                  focusable="false"
                  viewBox="0 0 1000 1000"
                >
                  <path
                    d="M694 242l249 250c12 11 12 21 1 32L694 773c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210-210H68c-13 0-23-10-23-23s10-23 23-23h806L662 275c-21-22 11-54 32-33z"
                  />
                </svg>
              </div>
              <div
                class="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-end-date"
                  aria-label="End Date"
                  autocomplete="off"
                  class="DateInput_input DateInput_input_1"
                  id="end-date"
                  name="end-date"
                  placeholder="End Date"
                  type="text"
                  value="04/14/2022"
                />
                <p
                  class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-end-date"
                >
                  Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-2 col-sm-12"
      >
        <a
          href="/?fromDate_=Invalid date&toDate_=2022-04-14"
        >
          <button
            class="m-1 btn btn-success"
            type="button"
          >
            OKE
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateRange renders correctly 1`] = `
<div>
  <div
    class="dateRange d-flex p-2"
  >
    <div
      class="row"
    >
      <div
        class="col-md-2 col-sm-12"
      >
        <div
          class="dateRange-title"
        >
          Range
        </div>
      </div>
      <div
        class="col-md-8 col-sm-12"
      >
        <div
          class="DateRangePicker DateRangePicker_1"
        >
          <div>
            <div
              class="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__withBorder DateRangePickerInput__withBorder_2"
            >
              <div
                class="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-start-date"
                  aria-label="Start Date"
                  autocomplete="off"
                  class="DateInput_input DateInput_input_1"
                  id="start-date"
                  name="start-date"
                  placeholder="Start Date"
                  type="text"
                  value="04/14/2022"
                />
                <p
                  class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-start-date"
                >
                  Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
              <div
                aria-hidden="true"
                class="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
                role="presentation"
              >
                <svg
                  class="DateRangePickerInput_arrow_svg DateRangePickerInput_arrow_svg_1"
                  focusable="false"
                  viewBox="0 0 1000 1000"
                >
                  <path
                    d="M694 242l249 250c12 11 12 21 1 32L694 773c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210-210H68c-13 0-23-10-23-23s10-23 23-23h806L662 275c-21-22 11-54 32-33z"
                  />
                </svg>
              </div>
              <div
                class="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-end-date"
                  aria-label="End Date"
                  autocomplete="off"
                  class="DateInput_input DateInput_input_1"
                  id="end-date"
                  name="end-date"
                  placeholder="End Date"
                  type="text"
                  value="04/14/2022"
                />
                <p
                  class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-end-date"
                >
                  Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-2 col-sm-12"
      >
        <a
          href="/?fromDate_=2022-04-14&toDate_=2022-04-14"
        >
          <button
            class="m-1 btn btn-success"
            type="button"
          >
            OKE
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
